deps/graph-parser/src/logseq/graph_parser/data_bridge/diff_merge.cljc

```diff
@@ -0,0 +1,154 @@
+                  (rest children))
+                 (concat
+                  (repeat (count next-children) (inc cur-level))
+                  (rest levels))))
+        result))))
+
+
+(defn db->diff-blocks
+  "db: datascript db
+   page-name: string"
+  [db page-name]
+  {:pre (string? page-name)}
+  (let [walked (get-sorted-page-block-ids db page-name)
+        blocks (d/pull-many db [:block/uuid :block/content :block/level] (map :id walked))
+        levels (map :level walked)
+        blocks (map (fn [block level]
+                      {:uuid   (str (:block/uuid block)) ;; Force to be string
+                       :body   (:block/content block)
+                       :level  level})
+                    blocks levels)]
+    blocks))
+
+;; Diverged from gp-block/extract-blocks
+;; From back to first to ensure end_pos is correct
+(defn ast->diff-blocks
+  "Prepare the blocks for diff-merge
+   blocks: ast of blocks
+   content: corresponding raw content"
+  [blocks content format {:keys [user-config block-pattern]}]
+  {:pre [(seq blocks) (string? content) (contains? #{:markdown :org} format)]}
+  (let [encoded-content (utf8/encode content)]
+    ;; (prn "DEBUG blocks:" blocks) ;; TODO Junyi
+    (loop [headings []
+           blocks (reverse blocks)
+           properties {}
+           end-pos (utf8/length encoded-content)]
+      (if (seq blocks)
+        (let [[block pos-meta] (first blocks)
+              ;; fix start_pos
+              pos-meta (assoc pos-meta :end_pos end-pos)]
+          (cond
+            (gp-block/heading-block? block)
+            (let [content (gp-block/get-block-content encoded-content block format pos-meta block-pattern)]
+              (recur (conj headings {:body  content
+                                     :level (:level (second block))
+                                     :uuid  (:id properties)})
+                     (rest blocks) {} (:start_pos pos-meta))) ;; The current block's start pos is the next block's end pos
+
+            (gp-property/properties-ast? block)
+            (let [new-props (:properties (gp-block/extract-properties (second block) (assoc user-config :format format)))]
+              ;; sending the current end pos to next, as it's not finished yet
+              ;; supports multiple properties sub-block possible in future
+              (recur headings (rest blocks) (merge properties new-props) (:end_pos pos-meta)))
+
+            :else
+            (recur headings (rest blocks) properties (:end_pos pos-meta))))
+        (if (empty? properties)
+          (reverse headings)
+          (let [[block _] (first blocks)
+                pos-meta {:start_pos 0 :end_pos end-pos}
+                content (gp-block/get-block-content encoded-content block format pos-meta block-pattern)
+                uuid (:id properties)]
+            (cons {:body content
+                   :level 1
+                   :uuid uuid}
+                  (reverse headings))))))))
```

Comment:
Here we construct blocks in the order of https://github.com/logseq/logseq/pull/8694/files#diff-59a4f1336962f1768912740687a6bee88d98d2df45d059e681e73b493e6aa15aR607